### PR TITLE
export va_list and ... correctly

### PR DESCRIPTION
**Pull request by dtcccc**

### CODE REVIEW

File: sched_boundary/sched_boundary.py [hunk 1]
```diff
@@ -154,27 +138,25 @@ def function_define(self, decl, _):
...
         elif obj in self.config['function']['fn_ptr']:
-            fn_ptr_export = fn_ptr_export_fmt.format(
-                fn=decl.name,
-                ret=GccBugs.fix(decl.result, decl.result.type.str_no_uid),
-                params=", ".join(GccBugs.fix(arg, arg.type.str_no_uid) \
+            signature = {"fn": decl.name,
+                "ret": GccBugs.fix(decl.result, decl.result.type.str_no_uid),
+                "params": ", ".join(GccBugs.fix(arg, arg.type.str_no_uid) \
                         for arg in decl.arguments) if decl.arguments else "void"
-            )
```
Comment (chenshanpei) [inline sched_boundary/sched_boundary.py:156]: It's better to store signature **string** in fn_ptr_list & interface_list.

File: sched_boundary/sched_boundary.py [hunk 2]
```diff
@@ -99,6 +99,11 @@ def fix(decl, str):
             str = bugfix(decl, str)
         return str
 
+    @staticmethod
+    def handle3dots(decl, signature):
```
Comment (chenshanpei) [inline sched_boundary/sched_boundary.py:103]: handle3dots ?  A better name is required.

File: sched_boundary/sched_boundary.py [hunk 3]
```diff
@@ -99,6 +99,11 @@ def fix(decl, str):
...
+        if decl.str_decl.find("...") >= 0:
+            signature["params"] += ", ..."
```
Comment (chenshanpei) [inline sched_boundary/sched_boundary.py:105]: try to use replace().

File: sched_boundary/sched_boundary.py [hunk 4]
```diff
@@ -143,6 +148,7 @@ def function_define(self, decl, _):
                 "params": ", ".join(GccBugs.fix(arg, arg.type.str_no_uid) \
                         for arg in decl.arguments) if decl.arguments else "void"
             }
+            GccBugs.handle3dots(decl, signature)
```
Comment (chenshanpei) [inline sched_boundary/sched_boundary.py:151]: It's better to export only one API (GccBugs.fix) from GccBugs class. I don't like the current design.

Comment (dtcccc) [reply sched_boundary/sched_boundary.py:151]: or even pack line 146-157

File: sched_boundary/sched_boundary.py (hunk 3 above)
Comment (chenshanpei) [reply sched_boundary/sched_boundary.py:105]: str_decl.find("...") >= 0 means there is "..." in str_decl. 
using str_del.replace("...",  ", ...") to keep the same as other code.

Comment (dtcccc) [reply sched_boundary/sched_boundary.py:105]: decl.str_decl is pure text from [[origin source kernel code]]. We can use it to check whether there is "..." but cannot directly use it.
The information extracted by gcc is stored in somewhere else. (e.g., return type is in decl.result, and params are in decl.arguments)

Here signature["params"] is generated by decl.arguments, and decl.arguments miss "..."

Comment (dtcccc) [reply sched_boundary/sched_boundary.py:105]: > decl.str_decl is pure text from [[origin kernel source code]]. 
Oh... there's a little wrong...
decl.str_decl is not 100% from origin kernel source code.

e.g., "enum" will miss, and "va_list" will become "struct  *" in decl.str_decl. (While "..." remains in it.)
So we could not use decl.str_decl when exporting to c files.

**Pull request merged**